File: app/src/main/java/workshop/akbolatss/xchangesrates/utils/logging/TimberLogImplementation.java
package workshop.akbolatss.xchangesrates.utils.logging;

import android.util.Log;

import com.google.firebase.crashlytics.FirebaseCrashlytics;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import timber.log.Timber;
import workshop.akbolatss.xchangesrates.BuildConfig;

public class TimberLogImplementation {

    public static void init() {
        if (BuildConfig.DEBUG)
            Timber.plant(new UmagPosDebugTree());
        else
            Timber.plant(new ProductionTree());
    }

    public static class UmagPosDebugTree extends Timber.DebugTree {
        @Override
        protected @Nullable String createStackElementTag(@NotNull StackTraceElement element) {
            return String.format("xChangesRates:C:%s:%s",
                    super.createStackElementTag(element),
                    element.getLineNumber());
        }

        @Override
        protected void log(int priority, @Nullable String tag, @NotNull String message, @Nullable Throwable t) {
            super.log(priority, tag, message, t);

            if (priority == Log.ERROR || priority == Log.WARN)
                if (t != null)
                    FirebaseCrashlytics.getInstance().recordException(t);
                else
                    FirebaseCrashlytics.getInstance().log(tag + ": Exception with `null` Throwable. Message " + message);
        }
    }

    public static class ProductionTree extends Timber.Tree {

        @Override
        protected void log(int priority, @Nullable String tag, @NotNull String message, @Nullable Throwable t) {
            if (priority == Log.ERROR || priority == Log.WARN)
                if (t != null)
                    FirebaseCrashlytics.getInstance().recordException(t);
                else
                    FirebaseCrashlytics.getInstance().log(tag + ": Exception with `null` Throwable. Message " + message);
        }
    }
}

File: app/src/main/java/workshop/akbolatss/xchangesrates/utils/logging/TimberLogImplementation.kt
package workshop.akbolatss.xchangesrates.utils.logging

import android.util.Log
import com.google.firebase.crashlytics.FirebaseCrashlytics
import timber.log.Timber
import timber.log.Timber.DebugTree
import workshop.akbolatss.xchangesrates.BuildConfig

object TimberLogImplementation {
    fun init() {
        if (BuildConfig.DEBUG) Timber.plant(DevelopmentTree()) else Timber.plant(
            ProductionTree()
        )
    }
}

private class DevelopmentTree : DebugTree() {
    override fun createStackElementTag(element: StackTraceElement): String? {
        return String.format(
            "xChangesRates:C:%s:%s",
            super.createStackElementTag(element),
            element.lineNumber
        )
    }

    override fun log(
        priority: Int,
        tag: String?,
        message: String,
        t: Throwable?
    ) {
        super.log(priority, tag, message, t)
        if (t != null && (priority == Log.ERROR || priority == Log.WARN))
            FirebaseCrashlytics.getInstance().recordException(t)
        else FirebaseCrashlytics.getInstance()
            .log("$tag: Exception with `null` Throwable. Message $message")
    }
}

private class ProductionTree : Timber.Tree() {
    override fun log(
        priority: Int,
        tag: String?,
        message: String,
        t: Throwable?
    ) {
        if (t != null && (priority == Log.ERROR || priority == Log.WARN))
            FirebaseCrashlytics.getInstance().recordException(t)
        else FirebaseCrashlytics.getInstance()
            .log("$tag: Exception with `null` Throwable. Message $message")
    }
}
